src/common-libs/plan/index.ts

```diff
@@ -88,38 +89,69 @@ const getAssociatedPlanVersion = async (
   return associatedPlanVersion;
 };
 
+/**
+ * Finds if any of the newly published reporting periods are higher
+ * (by looking at period number) then what's already stored as
+ * `lastPublishedReportingPeriodId`. Then, updates
+ * `planVersion.lastPublishedReportingPeriodId` if needed.
+ *
+ * Before actually updating the said field, this method checks
+ * whether measurements are already published for determined
+ * highest monitoring period.
+ */
 export const updateLastPublishedReportingPeriodId = async (
   models: Database,
-  planTag: InstanceDataOfModel<Database['planTag']>
+  planTag: InstanceDataOfModel<Database['planTag']>,
+  newlyPublishedReportingPeriodIds: PlanReportingPeriodId[]
 ) => {
+  const associatedPlanVersion = await getAssociatedPlanVersion(models, planTag);
+  const currentPublishedReportingPeriodId =
+    associatedPlanVersion.lastPublishedReportingPeriodId;
   const reportingPeriods = await models.planReportingPeriod.find({
     where: {
+      id: {
+        [models.Op.IN]: [
+          ...newlyPublishedReportingPeriodIds,
+          ...(currentPublishedReportingPeriodId
+            ? [createBrandedValue(currentPublishedReportingPeriodId)]
+            : []),
+        ],
+      },
       planId: planTag.planId,
       measurementsGenerated: true,
     },
-    orderBy: { column: 'periodNumber', order: 'desc' },
   });
 
   if (!reportingPeriods.length) {
     return;
   }
 
-  const latestReportingPeriodId = reportingPeriods[0].id;
+  let [latestPublishedReportingPeriod] = reportingPeriods;
+  const [, ...remainingPeriods] = reportingPeriods;
+  for (const reportingPeriod of remainingPeriods) {
+    if (
+      (reportingPeriod.periodNumber ?? -1) >
+      (latestPublishedReportingPeriod.periodNumber ?? -1)
+    ) {
+      latestPublishedReportingPeriod = reportingPeriod;
+    }
+  }
```

Comment:
We can replace for loop with below 
```suggestion
 const maxRemainingPeriod = remainingPeriods.reduce((p, c) =>
    p.periodNumber ?? -1 > (c.periodNumber ?? -1) ? p : c
  );
  if (
    (maxRemainingPeriod.periodNumber ?? -1) >
    (latestPublishedReportingPeriod.periodNumber ?? -1)
  ) {
    latestPublishedReportingPeriod = maxRemainingPeriod;
  }
```

Reply:
Actually, we can find `latestPublishedReportingPeriod` with a single `Array.prototype.reduce()` call and avoid `for..of` loop, but also avoid finding the first and `remainingPeriods`:
```typescript
const latestPublishedReportingPeriod = reportingPeriods.reduce((acc, curr) =>
  (curr.periodNumber ?? -1) > (acc.periodNumber ?? -1) ? curr : acc
);
```